File: backend/src/main/java/com/example/subjecthub/entity/Assessment.java
package com.example.subjecthub.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.annotation.ParametersAreNonnullByDefault;
import javax.persistence.*;

@Entity
@Table(name = "assessments")
@ParametersAreNonnullByDefault
public class Assessment {

    public enum AssessmentType {
        REPORT,
        TEST,
        FINAL,
        PROJECT
    }

    @Id
    @GeneratedValue
    @Column(nullable = false, name = "assessment_id")
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String description;

    @Column(nullable = false)
    private int weighting;

    @Column(nullable = false)
    private boolean groupWork;

    // Using a string here because length can differ e.g. could be "5000 words"
    // or "60 minutes" etc.
    @Column(nullable = false)
    private String length;

    @Column(nullable = false)
    private AssessmentType type;

    @JsonIgnore
    @ManyToOne
    @JoinColumn(name = "subject_id", nullable = false)
    private Subject subject;

    public Assessment() {
    }

    public Assessment(String name, String description, int weighting,
                      boolean groupWork, String length, AssessmentType type) {
        this.name = name;
        this.description = description;
        this.weighting = weighting;
        this.groupWork = groupWork;
        this.length = length;
        this.type = type;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getWeighting() {
        return weighting;
    }

    public void setWeighting(int weighting) {
        this.weighting = weighting;
    }

    public boolean isGroupWork() {
        return groupWork;
    }

    public void setGroupWork(boolean groupWork) {
        this.groupWork = groupWork;
    }

    public String getLength() {
        return length;
    }

    public void setLength(String length) {
        this.length = length;
    }

    public AssessmentType getType() { return type; }

    public void setType(AssessmentType type) { this.type = type; }

    public Subject getSubject() {
        return subject;
    }

    public void setSubject(Subject subject) {
        this.subject = subject;
    }

    @Override
    public String toString() {
        return "Assessment{" +
            " id=" + id +
            ", name='" + name + '\'' +
            ", type=" + type +
            ", subject=" + subject.getName() +
            '}';
    }
}

File: frontend/src/app/models/assessment.ts
export class Assessment {
  id: number;
  name: string;
  description: string;
  weighting: number;
  groupWork: boolean;
  length: string;
  type: AssessmentType;

  constructor(id?: number,
              name?: string,
              description?: string,
              weighting?: number,
              groupWork?: boolean,
              length?: string,
              type?: AssessmentType) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.weighting = weighting;
    this.groupWork = groupWork;
    this.length = length;
    this.type = type;
  }

  equals(other: any) {
    if (other == null || !(other instanceof Assessment)) {
      return false;
    }

    return this.id == other.id &&
      this.name == other.name &&
      this.description == other.description &&
      this.weighting == other.weighting &&
      this.groupWork == other.groupWork &&
      this.length == other.length &&
      this.type == other.type;
  }
}

// TODO: Upgrade to typescript 2.4 so we can use string enums
export enum AssessmentType {
  REPORT,
  TEST,
  FINAL,
  PROJECT
}

export function assessmentTypeFromString(s: string): AssessmentType {
  switch(s) {
    case 'PROJECT': return AssessmentType.PROJECT;
    case 'REPORT':  return AssessmentType.REPORT;
    case 'FINAL':   return AssessmentType.FINAL;
    case 'TEST':    return AssessmentType.TEST;
    default:
      throw new Error(`Unexpected assessment type: "${s}".`)
  }
}
